server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.query;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Represents a unique action that needs to be executed asynchronously during query rewriting.
+ * A {@link QueryRewriteAsyncAction} is registered using
+ * {@link QueryRewriteContext#registerUniqueRewriteAction(QueryRewriteAsyncAction, Consumer)}.
+ * This is useful when we want to remove duplicate and costly async actions that take part in query rewriting, such as generating
+ * embeddings for semantic search.
+ * Since we need to determine whether an action has already been registered, we require implementors to provide implementations for
+ * {@link #hashCode()} and {@link #equals(Object)},
+ */
+public abstract class QueryRewriteAsyncAction<T> {
+    private final QueryRewriteContext queryRewriteContext;
+
+    public QueryRewriteAsyncAction(QueryRewriteContext queryRewriteContext) {
+        this.queryRewriteContext = queryRewriteContext;
+    }
+
+    /**
+     * The execute method will:
+     * - Execute the action using {@link #execute(Client, ActionListener)}
+     * - Pass the action result to the consumers and execute the consumers
+     * - Call the final listener
+     *
+     * @param client A rest client that can be used during execution.
```

Comment:
Nit: internal client is different than rest client. I point it out because both exist.

```suggestion
     * @param client An internal client that can be used during execution.
```